src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -0,0 +1,2293 @@
+          > lhs.length=rhs->length
+          > lhs.data=*(rhs->data)
+
+\*******************************************************************/
+
+codet java_string_library_preprocesst::code_assign_java_string_to_string_expr(
+  const string_exprt &lhs, const exprt &rhs, symbol_tablet &symbol_table)
+{
+  assert(implements_java_char_sequence(rhs.type()));
+
+  typet deref_type;
+  if(rhs.type().subtype().id()==ID_symbol)
+    deref_type=symbol_table.lookup(
+      to_symbol_type(rhs.type().subtype()).get_identifier()).type;
+  else
+    deref_type=rhs.type().subtype();
+
+  dereference_exprt deref(rhs, deref_type);
+
+  // Fields of the string object
+  exprt rhs_length=get_length(deref, symbol_table);
+  exprt member_data=get_data(deref, symbol_table);
+  dereference_exprt rhs_data(member_data, member_data.type().subtype());
+
+  // Assignments
+  code_blockt code;
+  code.copy_to_operands(code_assignt(lhs.length(), rhs_length));
+  code.copy_to_operands(code_assignt(lhs.content(), rhs_data));
+  return code;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::
+            code_assign_string_literal_to_string_expr
+
+  Inputs:
+    lhs - an expression representing a java string
+    tmp_string - a temporary string to hold the literal
+    s - the literal to be assigned
+    symbol_table - symbol table
+
+  Output: return the following code:
+          > tmp_string = "s"
+          > lhs = (string_expr) tmp_string
+
+\*******************************************************************/
+
+codet java_string_library_preprocesst::
+  code_assign_string_literal_to_string_expr(
+    const string_exprt &lhs,
+    const exprt &tmp_string,
+    const std::string &s,
+    symbol_tablet &symbol_table)
+{
+  code_blockt code;
+  code.copy_to_operands(code_assignt(tmp_string, string_literal(s)));
+  code.copy_to_operands(code_assign_java_string_to_string_expr(
+    lhs, tmp_string, symbol_table));
+  return code;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::
+    make_string_builder_append_object_code
+
+  Inputs:
+    type - type of the function called
+    loc - location in the program
+    symbol_table - symbol table
+
+ Outputs: code for the StringBuilder.append(Object) function:
+          > string1 = arguments[1].toString()
+          > string_expr1 = string_to_string_expr(string1)
+          > string_expr2 = concat(this, string_expr1)
+          > this = string_expr_to_string(string_expr2)
+          > return this
+
+\*******************************************************************/
+
+codet java_string_library_preprocesst::make_string_builder_append_object_code(
+  const code_typet &type,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
+  code_typet::parameterst params=type.parameters();
+  exprt this_obj=symbol_exprt(params[0].get_identifier(), params[0].type());
```

Comment:
Yes, unless something goes really wrong in the interpretation of the bytecode. But I will protect the code by adding an assertion, which is always better than a segfault.